### PR TITLE
Add dependencies graph logic and update project structure

### PR DESCRIPTION
**Pull request by Hrom131**

### CODE REVIEW

File: src/internal/circom-files-cache.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
+import fsExtra from "fs-extra";
+import * as t from "io-ts";
+import * as path from "path";
+
+import { ProjectPathsConfig } from "hardhat/types";
+
+import { CIRCOM_CIRCUITS_CACHE_FILENAME } from "./constants";
+
+const FORMAT_VERSION = "hh-zkit-cache-1";
```
Comment (Arvolear) [inline src/internal/circom-files-cache.ts:9]: Mv to constants?

File: src/internal/constants.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+export const PLAGIN_NAME = "@solarity/hardhat-zkit";
+
+export const CIRCOM_CIRCUITS_CACHE_FILENAME = "circom-circuits-cache.json";
+
+export const INCLUDE_REG_EXP = /include +"([.\-/\w]+)";/g;
```
Comment (Arvolear) [inline src/internal/constants.ts:5]: Can it be single quoted?

File: src/internal/resolver.ts [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+    try {
+      packageJsonPath = this._resolveNodeModulesFileFromProjectRoot(path.join(libraryName, "package.json"));
+    } catch (error) {
+      // if the project is using a dependency from hardhat itself but it can't
+      // be found, this means that a global installation is being used, so we
+      // resolve the dependency relative to this file
+      if (libraryName === "hardhat") {
+        const hardhatCoreDir = path.join(__dirname, "..", "..");
+        packageJsonPath = path.join(hardhatCoreDir, "package.json");
+      } else {
```
Comment (Arvolear) [inline src/internal/resolver.ts:259]: Not needed

File: src/internal/resolver.ts [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+      await getRealPath(absolutePath),
+      libraryName,
+      libraryVersion,
+    );
+  }
+
+  private async _relativeImportToSourceName(from: ResolvedFile, imported: string): Promise<string> {
+    // This is a special case, were we turn relative imports from local files
+    // into library imports if necessary. The reason for this is that many
+    // users just do `import "../node_modules/lib/a.sol";`.
```
Comment (Arvolear) [inline src/internal/resolver.ts:314]:
```suggestion
    // users just do `import "../node_modules/lib/a.circom";`.
```

And other comments

File: src/internal/resolver.ts [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+import { createNonCryptographicHashBasedIdentifier } from "hardhat/internal/util/hash";
+
+import { Parser } from "./parse";
+import { CIRCOM_FILE_REG_EXP } from "./constants";
+
+export interface ResolvedFilesMap {
+  [sourceName: string]: ResolvedFile;
+}
+
+const NODE_MODULES = "node_modules";
```
Comment (Arvolear) [inline src/internal/resolver.ts:29]: Mv to const?

File: src/internal/dependencyGraph.ts [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    return [...transitiveDependencies];
+  }
+
+  public getConnectedComponents(): DependencyGraph[] {
+    const undirectedGraph: Record<string, Set<string>> = {};
+
+    for (const [sourceName, dependencies] of this._dependenciesPerFile.entries()) {
+      undirectedGraph[sourceName] = undirectedGraph[sourceName] ?? new Set();
+      for (const dependency of dependencies) {
+        undirectedGraph[dependency.sourceName] = undirectedGraph[dependency.sourceName] ?? new Set();
```
Comment (Arvolear) [inline src/internal/dependencyGraph.ts:67]: Please style up a bit

File: package.json [hunk 7]
```diff
@@ -56,18 +56,18 @@
   "dependencies": {
     "@nomicfoundation/hardhat-ethers": "3.0.5",
     "@solarity/zkit": "^0.1.0",
```
Comment (Arvolear) [inline package.json:58]: Strict deps please

File: src/internal/dependencyGraph.ts [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+  public getTransitiveDependencies(file: ResolvedFile): taskTypes.TransitiveDependency[] {
+    const visited = new Set<ResolvedFile>();
+
+    const transitiveDependencies = this._getTransitiveDependencies(file, visited, []);
+
+    return [...transitiveDependencies];
+  }
+
+  public getConnectedComponents(): DependencyGraph[] {
```
Comment (Arvolear) [inline src/internal/dependencyGraph.ts:61]: This function is not used. Do we need it?

File: package.json [hunk 9]
```diff
@@ -56,18 +56,18 @@
...
-    "ethers": "6.11.1"
-  },
-  "peerDependencies": {
-    "hardhat": "^2.10.0",
+    "ethers": "6.11.1",
+    "hardhat": "^2.20.0",
```
Comment (Arvolear) [inline package.json:60]: Hardhat has to be a peer dependency

File: package.json [hunk 10]
```diff
@@ -81,6 +81,7 @@
     "mocha": "^10.3.0",
     "nyc": "^15.1.0",
     "prettier": "^3.2.5",
+    "resolve": "1.22.8",
```
Comment (Arvolear) [inline package.json:84]: `resolve` has to be a dependency

File: package.json [hunk 11]
```diff
@@ -56,18 +56,18 @@
   "dependencies": {
     "@nomicfoundation/hardhat-ethers": "3.0.5",
     "@solarity/zkit": "^0.1.0",
-    "ethers": "6.11.1"
-  },
-  "peerDependencies": {
-    "hardhat": "^2.10.0",
+    "ethers": "6.11.1",
+    "hardhat": "^2.20.0",
     "typechain": "^8.0.0"
```
Comment (Arvolear) [inline package.json:61]: Why do you need typechain?

File: src/internal/dependencyGraph.ts
Comment (Arvolear) [inline src/internal/dependencyGraph.ts:1]: Usually if a file exports a class, a file name matches a class name.

File: package.json
Comment (Arvolear) [inline package.json:1]: "keywords": [
    "hardhat",
    "hardhat-plugin",
    "distributedlab",
    "solarity",
    "circom",
    "zk",
    "zkit"
  ],

File: src/tasks/compile.ts [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+        artifactsDir ?? config.zkit.compilationSettings.artifactsDir,
+      );
+      circuitFilesCache = await invalidateCacheMissingArtifacts(
+        circuitsRoot,
+        artifactsDirFullPath,
+        circuitFilesCache,
+        resolvedFilesToCompile,
+      );
+
+      const resolvedFilesWithDependecies: ResolvedFileWithDependencies[] = await run(
```
Comment (Arvolear) [inline src/tasks/compile.ts:255]:
```suggestion
      const resolvedFilesWithDependencies: ResolvedFileWithDependencies[] = await run(
```

File: src/internal/resolver.ts [hunk 13]
```diff
@@ -0,0 +1,461 @@
...
+    return (
+      this._isRelativeImport(from, imported) && from.library === undefined && imported.includes(`${NODE_MODULES}/`)
+    );
+  }
+
+  private _relativeImportToLibraryToSourceName(from: ResolvedFile, imported: string): string {
+    const sourceName = normalizeSourceName(path.join(path.dirname(from.sourceName), imported));
+
+    const nmIndex = sourceName.indexOf(`${NODE_MODULES}/`);
+    return sourceName.substr(nmIndex + NODE_MODULES.length + 1);
```
Comment (Arvolear) [inline src/internal/resolver.ts:432]: `substr` is deprecated

```suggestion
    return sourceName.substring(nmIndex + NODE_MODULES.length + 1);
```

File: src/internal/resolver.ts [hunk 14]
```diff
@@ -0,0 +1,461 @@
...
+
+  private async _resolveLocalSourceName(sourceName: string, remappedSourceName: string): Promise<ResolvedFile> {
+    await this._validateSourceNameExistenceAndCasing(this._projectRoot, remappedSourceName, false);
+
+    const absolutePath = path.join(this._projectRoot, remappedSourceName);
+    return this._resolveFile(sourceName, absolutePath);
+  }
+
+  private async _resolveLibrarySourceName(sourceName: string, remappedSourceName: string): Promise<ResolvedFile> {
+    const normalizedSourceName = remappedSourceName.replace(/^node_modules\//, "");
```
Comment (Arvolear) [inline src/internal/resolver.ts:246]: Can we use a constant here?

File: src/internal/constants.ts (hunk 2 above)
Comment (Hrom131) [reply src/internal/constants.ts:5]: No, only double quotes are supported by the compiler

File: src/internal/dependencyGraph.ts (hunk 8 above)
Comment (Hrom131) [reply src/internal/dependencyGraph.ts]: No, we don't need this functionality in the current version, so I removed this function

File: package.json (hunk 11 above)
Comment (Hrom131) [reply package.json]: We need typechain only for tests

File: src/internal/resolver.ts (hunk 14 above)
Comment (Hrom131) [reply src/internal/resolver.ts]: Yes, changed it to use a constant

File: package.json [hunk 15]
```diff
@@ -56,18 +56,18 @@
   "dependencies": {
     "@nomicfoundation/hardhat-ethers": "3.0.5",
```
[the remark this answers is not in the document]
Comment (Hrom131) [reply package.json]: We don't need this plugin in the dependencies, so moved it to dev dependencies

**Pull request merged**